File: dbflow-tests/src/test/java/com/raizlabs/android/dbflow/test/sql/MigrationTest.java
package com.raizlabs.android.dbflow.test.sql;

import android.database.Cursor;
import android.database.sqlite.SQLiteException;

import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.sql.SQLiteType;
import com.raizlabs.android.dbflow.sql.language.Select;
import com.raizlabs.android.dbflow.sql.migration.AlterTableMigration;
import com.raizlabs.android.dbflow.sql.migration.IndexMigration;
import com.raizlabs.android.dbflow.sql.migration.UpdateTableMigration;
import com.raizlabs.android.dbflow.test.FlowTestCase;

import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class MigrationTest extends FlowTestCase {

    @Test
    public void testMigration() {

        List<String> columnNames = Arrays.asList("`fraction` REAL", "`time` INTEGER", "`name2` TEXT", "`number` INTEGER", "`blobby` BLOB");
        List<String> columns = Arrays.asList("fraction", "time", "name2", "number", "blobby");

        AlterTableMigration<MigrationModel> renameMigration = new AlterTableMigration<>(MigrationModel.class).renameFrom("TestModel");
        renameMigration.onPreMigrate();
        assertEquals("ALTER TABLE `TestModel` RENAME TO `MigrationModel`", renameMigration.getRenameQuery());
        renameMigration.onPostMigrate();

        AlterTableMigration<MigrationModel> alterTableMigration = new AlterTableMigration<>(MigrationModel.class);
        alterTableMigration.addColumn(SQLiteType.REAL, "fraction")
                .addColumn(SQLiteType.INTEGER, "time")
                .addColumn(SQLiteType.TEXT, "name2")
                .addColumn(SQLiteType.INTEGER, "number")
                .addColumn(SQLiteType.BLOB, "blobby");
        alterTableMigration.onPreMigrate();

        List<String> columnDefinitions = alterTableMigration.getColumnDefinitions();
        for (int i = 0; i < columnDefinitions.size(); i++) {
            assertEquals("ALTER TABLE `MigrationModel` ADD COLUMN " + columnNames.get(i), columnDefinitions.get(i));
        }

        try {
            alterTableMigration.migrate(FlowManager.getDatabaseForTable(MigrationModel.class).getWritableDatabase());
        } catch (SQLiteException e) {
            if (e.getMessage().startsWith("duplicate column name: fraction (code 1):")) {
                // ignore since we've already added the column.
            } else {
                // some other issue
                throw new RuntimeException(e);
            }
        }

        // test the column sizes
        Cursor cursor = new Select().from(MigrationModel.class).where().query();
        assertTrue(cursor.getColumnNames().length == columnNames.size() + 2);

        // make sure column exists now
        for (int i = 0; i < columns.size(); i++) {
            assertTrue(cursor.getColumnIndex(columns.get(i)) != -1);
        }

        alterTableMigration.onPostMigrate();
    }

    @Test
    public void testUpdateMigration() {
        UpdateTableMigration<MigrationModel> updateTableMigration
                = new UpdateTableMigration<>(MigrationModel.class)
                .set(MigrationModel_Table.name.is("test")).where(MigrationModel_Table.name.is("notTest"));
        updateTableMigration.onPreMigrate();

        assertEquals("UPDATE `MigrationModel` SET `name`='test' WHERE `name`='notTest'", updateTableMigration.getQuery().trim());

        updateTableMigration.migrate(FlowManager.getDatabaseForTable(MigrationModel.class).getWritableDatabase());
        updateTableMigration.onPostMigrate();
    }

    @Test
    public void testSqlFile() {
        MigrationModel migrationModel = new MigrationModel();
        migrationModel.setName("test");
        migrationModel.save();
        Cursor cursor = new Select().from(MigrationModel.class).query();
        assertTrue(cursor.moveToFirst());

        int addedColumIndex = cursor.getColumnIndex("addedColumn");
        assertFalse(addedColumIndex == -1);
    }

    public void testIndexMigration() {
        IndexMigration<TestModel3> indexMigration
                = new IndexMigration<>("MyIndex", TestModel3.class)
                .addColumn(TestModel3_Table.type);
        assertEquals("CREATE INDEX IF NOT EXISTS `MyIndex` ON `TestModel32`(`type`)", indexMigration.getIndexQuery().trim());
    }

}

File: dbflow-tests/src/test/java/com/raizlabs/android/dbflow/test/sql/MigrationTest.kt
package com.raizlabs.android.dbflow.test.sql

import android.database.Cursor

import com.raizlabs.android.dbflow.config.FlowManager
import com.raizlabs.android.dbflow.sql.SQLiteType
import com.raizlabs.android.dbflow.sql.language.Select
import com.raizlabs.android.dbflow.sql.migration.AlterTableMigration
import com.raizlabs.android.dbflow.sql.migration.IndexMigration
import com.raizlabs.android.dbflow.sql.migration.UpdateTableMigration
import com.raizlabs.android.dbflow.test.FlowTestCase

import org.junit.Test

import java.util.Arrays

import junit.framework.Assert.assertFalse
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue

class MigrationTest : FlowTestCase() {

    @Test
    fun testMigration() {

        val columnNames = Arrays.asList("`fraction` REAL", "`time` INTEGER", "`name2` TEXT", "`number` INTEGER", "`blobby` BLOB")
        val columns = Arrays.asList("fraction", "time", "name2", "number", "blobby")

        val renameMigration = AlterTableMigration(MigrationModel::class.java).renameFrom("TestModel")
        renameMigration.onPreMigrate()
        assertEquals("ALTER TABLE `TestModel` RENAME TO `MigrationModel`", renameMigration.renameQuery)
        renameMigration.onPostMigrate()

        val alterTableMigration = AlterTableMigration(MigrationModel::class.java)
        alterTableMigration.addColumn(SQLiteType.REAL, "fraction")
                .addColumn(SQLiteType.INTEGER, "time")
                .addColumn(SQLiteType.TEXT, "name2")
                .addColumn(SQLiteType.INTEGER, "number")
                .addColumn(SQLiteType.BLOB, "blobby")
        alterTableMigration.onPreMigrate()

        val columnDefinitions = alterTableMigration.columnDefinitions
        for (i in columnDefinitions.indices) {
            assertEquals("ALTER TABLE `MigrationModel` ADD COLUMN " + columnNames[i], columnDefinitions[i])
        }

        alterTableMigration.migrate(FlowManager.getDatabaseForTable(MigrationModel::class.java).writableDatabase)

        // test the column sizes
        val cursor = Select().from(MigrationModel::class.java).where().query()
        assertTrue(cursor!!.columnNames.size == columnNames.size + 2)

        try {
            Thread.sleep(200)
        } catch (e: InterruptedException) {
        }

        // make sure column exists now
        for (i in columns.indices) {
            assertTrue(cursor.getColumnIndex(columns[i]) != -1)
        }
        cursor.close()

        alterTableMigration.onPostMigrate()
    }

    @Test
    fun testUpdateMigration() {
        val updateTableMigration = UpdateTableMigration(MigrationModel::class.java)
                .set(MigrationModel_Table.name.`is`("test")).where(MigrationModel_Table.name.`is`("notTest"))
        updateTableMigration.onPreMigrate()

        assertEquals("UPDATE `MigrationModel` SET `name`='test' WHERE `name`='notTest'", updateTableMigration
                .updateStatement.query.trim { it <= ' ' })

        updateTableMigration.migrate(FlowManager.getDatabaseForTable(MigrationModel::class.java).writableDatabase)
        updateTableMigration.onPostMigrate()
    }

    @Test
    fun testSqlFile() {
        val migrationModel = MigrationModel()
        migrationModel.name = "test"
        migrationModel.save()
        val cursor = Select().from(MigrationModel::class.java).query()
        assertTrue(cursor!!.moveToFirst())

        val addedColumIndex = cursor.getColumnIndex("addedColumn")
        assertFalse(addedColumIndex == -1)

        cursor.close()
        // broken with junit tests
    }

    fun testIndexMigration() {
        val indexMigration = object : IndexMigration<TestModel3>(TestModel3::class.java) {
            override fun getName(): String {
                return "MyIndex"
            }
        }
                .addColumn(TestModel3_Table.type)
        assertEquals("CREATE INDEX IF NOT EXISTS `MyIndex` ON `TestModel32`(`type`)", indexMigration.indexQuery.trim { it <= ' ' })
    }

}
